### ISSUE TITLE
socket.io transport error "ECONNRESET"

### ISSUE THREAD

**Issue by stanleytakamatsu** [socketio/socket.io#1647]
When I use socket.io-client in a browser, it connects normally. 
But when I use from nodejs server, it generates a connection error.

engine.io-client:socket socket error {"description":{"code":"ECONNRESET","target":{"_socket":null,"bytesReceived":0,"readyState":0,"supports":{"binary":true},"_isServer":false,"url":"wss://valid.address/socket.io/?EIO=3&transport=websocket","protocolVersion":13,"binaryType":"arraybuffer","_events":{}}}}

Someone could help me, please.
thanks

**Comment by rase-** [comment]
Does this still happen?

**Comment by thrar** [comment]
I'm getting ECONNRESET as well when connecting from NodeJS. Connecting from a browser is fine.

The error looks like this:

  { [Error: websocket error]
    type: 'TransportError',
    description: 
   { [Error: socket hang up]
     code: 'ECONNRESET',
     target: 
      { _socket: null,
        bytesReceived: 0,
        readyState: 0,
        supports: [Object],
        _isServer: false,
        url: 'ws://localhost:9000/socket.io/?EIO=3&transport=websocket',
        protocolVersion: 13,
        binaryType: 'arraybuffer',
        _events: [Object] } } }

**Comment by thrar** [comment]
Figured it out. I had to set a path in my socket.io-client options, like so:

  options = {
    transports: ['websocket'],
    'force new connection': true,
    path: '/socket.io-client'
  }

This works fine for me.

**Comment by uiteoi** [comment]
I have the same issue using node.js client with socket.io clent and server version 1.3.7 (latest at this time). I have tried to setup a path as mentioned in the previous comment to no avail. I had no issues getting this to work with the same code using version 0.9.16. This is the last (of many) issues preventing me to upgrade to version 0.9.16, now seriously considering non-socket.io options.

Last client code tried:
```
var io = require( 'socket.io-client' ); // version 1.3.7
io.connect( 'localhost:8080', { transports: [ 'websocket' ], path: "/socket.io/socket.io.js" } )
```

Last server code tried:
```
var socketio = require( 'socket.io' ); // version 1.3.7
var io = socketio.listen( server );
io.use( function( socket, next ) { next() } );
io.sockets.on( 'connection', function( socket ){ ... } )
```


Here is the full trace:
```
Mon, 30 Nov 2015 09:42:19 GMT socket.io-client:manager opening https://localhost:8080
Mon, 30 Nov 2015 09:42:19 GMT engine.io-client:socket creating transport "websocket"
Mon, 30 Nov 2015 09:42:19 GMT socket.io-client:manager connect attempt will timeout after 20000
Mon, 30 Nov 2015 09:42:19 GMT engine.io-client:socket setting transport websocket
Mon, 30 Nov 2015 09:42:19 GMT socket.io-client:manager readyState opening
Mon, 30 Nov 2015 09:42:20 GMT engine.io-client:socket socket error {"description":{"code":"ECONNRESET","target":{"domain":null,"_events":{},"_maxListeners":10,"_socket":null,"_ultron":null,"_closeReceived":false,"bytesReceived":0,"readyState":0,"supports":{"binary":true},"extensions":{},"_isServer":false,"url":"wss://localhost:8080/socket.io/socket.io.js/?EIO=3&transport=websocket","protocolVersion":13,"binaryType":"arraybuffer"}}}
Mon, 30 Nov 2015 09:42:20 GMT socket.io-client:manager connect_error
Mon, 30 Nov 2015 09:42:20 GMT engine.io-client:socket socket close with reason: "transport error"
Mon, 30 Nov 2015 09:42:20 GMT socket.io-client:manager will wait 755ms before reconnect attempt
Mon, 30 Nov 2015 09:42:21 GMT socket.io-client:manager attempting reconnect
Mon, 30 Nov 2015 09:42:21 GMT engine.io-client:socket creating transport "websocket"
Mon, 30 Nov 2015 09:42:21 GMT socket.io-client:manager readyState closed
Mon, 30 Nov 2015 09:42:21 GMT engine.io-client:socket setting transport websocket
Mon, 30 Nov 2015 09:42:21 GMT socket.io-client:manager opening https://localhost:8080
Mon, 30 Nov 2015 09:42:21 GMT engine.io-client:socket socket error {"description":{"code":"ECONNRESET","target":{"domain":null,"_events":{},"_maxListeners":10,"_socket":null,"_ultron":null,"_closeReceived":false,"bytesReceived":0,"readyState":0,"supports":{"binary":true},"extensions":{},"_isServer":false,"url":"wss://localhost:8080/socket.io/socket.io.js/?EIO=3&transport=websocket","protocolVersion":13,"binaryType":"arraybuffer"}}}
Mon, 30 Nov 2015 09:42:21 GMT socket.io-client:manager connect attempt will timeout after 20000
... and so on
```

**Comment by uiteoi** [comment]
My problem in this test came from the specified location, I did not specify a protocol and socket.io assumed that it was "https" when my test server is "http" only.

Using the following client code fixed my problem, there is no need to specify a path, or transports although specifying ["websocket"] also works as shown:
```
var io = require( 'socket.io-client' ); // version 1.3.7
io.connect( 'http://localhost:8080', { transports: [ 'websocket' ] } )
```

When not specifying transports, socket.io uses both xhr and websocket as the web client does.

**Comment by efkan** [comment]
I get this error when trying to connect server's own as a client.
I mean I want to run Socket.io and Socket.io-client on the same machine.

**Comment by efkan** [comment]
I've overcome this issue by splitting the Express http server and socket.io server.

**Comment by darrachequesne** [comment]
That issue was closed automatically. Please check if your issue is fixed with the latest release, and reopen if needed (with a [fiddle](https://github.com/darrachequesne/socket.io-fiddle) reproducing the issue if possible).

**Comment by Loilo** [comment]
Maybe a hint for people googling this:

I had this issue because I used an SSL server for socket.io, but still connected to `localhost`. This still points to the correct server IP, however, since the certificate was not issued for the `localhost` domain, socket.io (or rather the underlying Node.js connection I suppose) was confused and failed to connect.

I had to specifically connect to the domain the certificate was issued for, then my problems were gone.

```javascript
// Say the certificate was issued for example.com
require('socket.io-client').connect('https://example.com:8080')
```
